WARC spout to emit captures into topology (implements #755)

sebastian-nagel opened the pull request:
A simple WARC spout:
- inherits from FileSpout to read paths/URLs from local files
- based on [jwarc](//github.com/iipc/jwarc)
- mimics the behavior of FetcherBolt, i.e.
  - all captures are sent to status stream
  - only successful fetches (HTTP status 200) will emit content to default stream for parsing, indexing, etc.

Tested so far only with a few WARC files, both local and Common Crawl files referenced by URL `http://commoncrawl.s3.amazonaws.com/crawl-data/...`

jnioche commented:
Nice one @sebastian-nagel! 

Would some of the existing components of the WARC module benefit from **jwarc**?

Our code to generate WARCs leverages **storm-hdfs** (and can of course deal with local files as well), why not have a spout based on HdfsSpout instead? This way we could deal with remote and or larger input data?

sebastian-nagel commented:
> Would some of the existing components of the WARC module benefit from **jwarc**?

Maybe, it's on my list to try to rewrite the WARCHdfsBolt using jwarc.

> why not have a spout based on HdfsSpout instead?

After a look on [HDFS spout](https://github.com/apache/storm/blob/master/docs/storm-hdfs.md#hdfs-spout): the way files are marked before and moved after processing seems not really applicable for web archives - those usually stay in place. We would need the same mechanism as in the first version of WarcSpout: write paths/URLs pointing to WARCs into files and put these in the input folder of HdfsSpout. I'll think about it. Maybe it's the next step.

jnioche commented:
makes sense

Could you please update the README for the WARC module as part of the PR? Good to go otherwise

thanks

sebastian-nagel commented:
Yes, I'll update the README. Give me a few days to do some more testing - I've found already one open point: WARC records with payload still compressed using Content-Encoding.

sebastian-nagel commented:
Rebased to current master and addressed the following points:
- added info about warc spout to README
- store WARC record location to metadata (warc.file.name, warc.record.offset and warc.record.length), useful for debugging if the content caused a parser failure, related to #567 
- upgrade jwarc dependency to 0.12.0
- add configurable sleep after fetches to work around OOMs due to large overflow queues

Open points:
- HTTP Content-Encoding is not decoded. Waiting whether a method for this could be made available in jwarc, see iipc/jwarc#48
- a single WARC spout is able to emit 2,000-3,000 "fetches" per sec. which makes any topology consuming the content reliably fail by an OOM caused by >20k tuples in the overflow queue. I've tried without success: shrink `topology.max.spout.pending`, backpressure enabled/disabled and various values for topology.message.timeout.secs (30-300 sec.)


Testing:
- "dummy" topology in local mode: WARCSpout ( -> DummyIndexer ) -> StdOutStatusUpdater
  - processed 70 WARC files
  - average: 2,400 WARC records per sec.
  - no sleep
- cluster mode, with 10 ms. sleep after "fetches":
  - rewrite WARC files: WARCSpout -> WARCHdfsBolt
  - parse and index content

jnioche commented on external/warc/src/main/java/com/digitalpebble/stormcrawler/warc/WARCSpout.java:
```diff
@@ -0,0 +1,530 @@
+        if (warcReader == null && buffer.isEmpty()) {
+            // input exhausted
+            try {
+                Thread.sleep(10);
```
don't need it -  Storm will handle sleeping between calls to nextTuple if you don't emit anything during a call

sebastian-nagel replied:
Ok, will remove it - was following the API doc of [ISpout::nextTuple](https://storm.apache.org/releases/1.2.3/javadocs/org/apache/storm/spout/ISpout.html#nextTuple--), obviously outdated.

jnioche replied:
Let me double check in Storm's code

jnioche replied:
see http://storm.apache.org/releases/current/Performance.html -> wait strategy

default values 

_topology.sleep.spout.wait.strategy.time.ms | 1
topology.spout.wait.strategy | "org.apache.storm.spout.SleepSpoutWaitStrategy"_

sebastian-nagel replied:
Ok: sleep removed.

jnioche commented on external/warc/src/main/java/com/digitalpebble/stormcrawler/warc/WARCSpout.java:
```diff
@@ -0,0 +1,530 @@
+            return;
+
+        LOG.info("Reading WARC file {}", warcFileInProgress);
+        ReadableByteChannel warcChannel = null;
```
try with resource to simplify the code and remove the close section below?

sebastian-nagel replied:
Good point. Thanks!

sebastian-nagel replied:
Actually, it does not work: the channel needs to stay, it'll be closed together by `warcReader.close()`.

jnioche replied:
ok!

jnioche commented on external/warc/src/main/java/com/digitalpebble/stormcrawler/warc/WARCSpout.java:
```diff
@@ -0,0 +1,530 @@
+                LOG.debug("Skipped WARC record of type {}",
+                        record.get().type());
+            }
+            if (storeHTTPHeaders && "request".equals(warcType)) {
```
that string "request" is used in other places in the module - have it as a constant of enum somewhere?

sebastian-nagel replied:
Ok. But maybe better use `instanceof WarcRequest` as the literal "request" is an internal of jwarc.

jnioche commented:
Implementing our own timeout mechanism is not right, max spout pending should be the way to go.

Is there more than one _emit_ per call to _nextTuple()_? if so this would explain why _topology.max.spout.pending_ doesn't have any effect. 

Looking at the code this does not seem to be the case though, any idea?

sebastian-nagel commented:
Hi @jnioche, I've addressed all your comments. Some example topologies which push WARC content into the topology are in [sebastian-nagel/warc-crawler](/sebastian-nagel/warc-crawler). The "devnull" topology reads 1.2 million WARC records in 9 min. on my laptop. However, the workers of the "rewarc" topo repeatedly crash with an OOM exception. An analysis of the heap dump shows that almost the entire memory is occupied by the overflow queue (cf. [here](http://mail-archives.apache.org/mod_mbox/storm-user/201705.mbox/%3C913EBAD0-2431-469E-B1D9-B2E07DFF9725@hortonworks.com%3E)). Some 20k tuples are emitted but not yet acked:
![Screenshot_2020-06-23StormUI-rewarc](https://user-images.githubusercontent.com/1630582/85463192-a5dacd00-b5a6-11ea-8312-bd5fcfeb070d.png)
The WARCHdfsBolt is considerably slow (checksums and gzip are computationally expensive), but looks like that max. pending is not applied (maybe due to a misconfiguration).

jnioche commented:
Thanks @sebastian-nagel I can reproduce the issue and will look into it.

jnioche commented:
You need to emit with a tuple ID, otherwise, Storm won't track the acks and fails and the number of tuples being processed.

`            _collector.emit(new Values(url, content, metadata), url);
`
you might need to set   `topology.backpressure.enable: false` for `max.spout.pending` to be effective

The change above seems to have fixed the issue in my case.

jnioche commented on external/warc/src/main/java/com/digitalpebble/stormcrawler/warc/WARCSpout.java:
```diff
@@ -0,0 +1,520 @@
+
+    private int maxContentSize = -1;
+    private int contentBufferSize = 8192;
+    private long sleepEmitFetched = 0;
```
not used - remove?

sebastian-nagel replied:
Done.

sebastian-nagel commented:
Thanks, @jnioche! I've added the missing ID to the emit call. I've run only a short test: looks good, memory consumption is low.

jnioche commented:
Thanks @sebastian-nagel, this is a great addition to our WARC module.
